Requests should not be redirected if there is no scheduling service

## Bug Report

```
[2023/12/08 07:01:19.880 +00:00] [ERROR] [restore.go:59] ["failed to restore"] [error="[domain:8243][PD:apiutil:ErrRedirect]redirect failed\n"] [errorVerbose="[domain:8243][PD:apiutil:ErrRedirect]redirect failed\n\ngithub.com/pingcap/errors.AddStack\n\t/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174\ngithub.com/pingcap/errors.Trace\n\t/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/juju_adaptor.go:15\ngithub.com/pingcap/tidb/ddl.(*ddl).DoDDLJob\n\t/go/src/github.com/pingcap/tidb/ddl/ddl.go:1176\ngithub.com/pingcap/tidb/ddl.(*ddl).BatchCreateTableWithInfo\n\t/go/src/github.com/pingcap/tidb/ddl/ddl_api.go:2645\ngithub.com/pingcap/tidb/br/pkg/gluetidb.(*tidbSession).SplitBatchCreateTable\n\t/go/src/github.com/pingcap/tidb/br/pkg/gluetidb/glue.go:225\ngithub.com/pingcap/tidb/br/pkg/gluetidb.(*tidbSession).CreateTables\n\t/go/src/github.com/pingcap/tidb/br/pkg/gluetidb/glue.go:273\ngithub.com/pingcap/tidb/br/pkg/restore.(*DB).CreateTables\n\t/go/src/github.com/pingcap/tidb/br/pkg/restore/db.go:316\ngithub.com/pingcap/tidb/br/pkg/restore.(*Client).createTables\n\t/go/src/github.com/pingcap/tidb/br/pkg/restore/client.go:697\ngithub.com/pingcap/tidb/br/pkg/restore.(*Client).createTablesInWorkerPool.func1\n\t/go/src/github.com/pingcap/tidb/br/pkg/restore/client.go:889\ngithub.com/pingcap/tidb/br/pkg/utils.(*WorkerPool).ApplyWithIDInErrorGroup.func1\n\t/go/src/github.com/pingcap/tidb/br/pkg/utils/worker.go:85\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\t/go/pkg/mod/golang.org/x/sync@v0.2.0/errgroup/errgroup.go:75\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1594"] [stack="main.runRestoreCommand\n\t/go/src/github.com/pingcap/tidb/br/cmd/br/restore.go:59\nmain.newDBRestoreCommand.func1\n\t/go/src/github.com/pingcap/tidb/br/cmd/br/restore.go:157\ngithub.com/spf13/cobra.(*Command).execute\n\t/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:916\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:1044\ngithub.com/spf13/cobra.(*Command).Execute\n\t/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:968\nmain.main\n\t/go/src/github.com/pingcap/tidb/br/cmd/br/main.go:57\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:250"]
Error: [domain:8243][PD:apiutil:ErrRedirect]redirect failed

[2023/12/08 07:01:19.880 +00:00] [ERROR] [main.go:59] ["br failed"] [error="[domain:8243][PD:apiutil:ErrRedirect]redirect failed\n"] [errorVerbose="[domain:8243][PD:apiutil:ErrRedirect]redirect failed\n\ngithub.com/pingcap/errors.AddStack\n\t/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174\ngithub.com/pingcap/errors.Trace\n\t/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/juju_adaptor.go:15\ngithub.com/pingcap/tidb/ddl.(*ddl).DoDDLJob\n\t/go/src/github.com/pingcap/tidb/ddl/ddl.go:1176\ngithub.com/pingcap/tidb/ddl.(*ddl).BatchCreateTableWithInfo\n\t/go/src/github.com/pingcap/tidb/ddl/ddl_api.go:2645\ngithub.com/pingcap/tidb/br/pkg/gluetidb.(*tidbSession).SplitBatchCreateTable\n\t/go/src/github.com/pingcap/tidb/br/pkg/gluetidb/glue.go:225\ngithub.com/pingcap/tidb/br/pkg/gluetidb.(*tidbSession).CreateTables\n\t/go/src/github.com/pingcap/tidb/br/pkg/gluetidb/glue.go:273\ngithub.com/pingcap/tidb/br/pkg/restore.(*DB).CreateTables\n\t/go/src/github.com/pingcap/tidb/br/pkg/restore/db.go:316\ngithub.com/pingcap/tidb/br/pkg/restore.(*Client).createTables\n\t/go/src/github.com/pingcap/tidb/br/pkg/restore/client.go:697\ngithub.com/pingcap/tidb/br/pkg/restore.(*Client).createTablesInWorkerPool.func1\n\t/go/src/github.com/pingcap/tidb/br/pkg/restore/client.go:889\ngithub.com/pingcap/tidb/br/pkg/utils.(*WorkerPool).ApplyWithIDInErrorGroup.func1\n\t/go/src/github.com/pingcap/tidb/br/pkg/utils/worker.go:85\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\t/go/pkg/mod/golang.org/x/sync@v0.2.0/errgroup/errgroup.go:75\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1594"] [stack="main.main\n\t/go/src/github.com/pingcap/tidb/br/cmd/br/main.go:59\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:250"]
```

<img width="1745" alt="Screenshot 2023-12-08 at 16 42 42" src="https://github.com/tikv/pd/assets/35896542/c6810706-958b-4652-816f-13998a69384e">

But actually, there is no scheduling service, the placement rule request should not be forwarded to the scheduling service. Here is the related code:
https://github.com/tikv/pd/blob/ab3025c4d56dc4e89812db7d227466b906fe2a49/pkg/utils/apiutil/serverapi/middleware.go#L114-L160

--- a/pkg/mcs/scheduling/server/config/config.go
+++ b/pkg/mcs/scheduling/server/config/config.go
@@ -612,7 +612,7 @@ func (o *PersistConfig) IsLocationReplacementEnabled() bool {
 	return o.getTTLBoolOr(sc.EnableLocationReplacement, o.GetScheduleConfig().EnableLocationReplacement)
 }
 
-// IsTikvRegionSplitEnabled returns whether tikv split region is disabled.
+// IsTikvRegionSplitEnabled returns whether tikv split region is enabled.
 func (o *PersistConfig) IsTikvRegionSplitEnabled() bool {
 	return o.getTTLBoolOr(sc.EnableTiKVSplitRegion, o.GetScheduleConfig().EnableTiKVSplitRegion)
 }

--- a/pkg/utils/apiutil/serverapi/middleware.go
+++ b/pkg/utils/apiutil/serverapi/middleware.go
@@ -22,6 +22,7 @@ import (
 	"github.com/pingcap/failpoint"
 	"github.com/pingcap/log"
 	"github.com/tikv/pd/pkg/errs"
+	mcsutils "github.com/tikv/pd/pkg/mcs/utils"
 	"github.com/tikv/pd/pkg/slice"
 	"github.com/tikv/pd/pkg/utils/apiutil"
 	"github.com/tikv/pd/server"
@@ -122,16 +123,24 @@ func (h *redirector) matchMicroServiceRedirectRules(r *http.Request) (bool, stri
 	// It will be helpful when matching the redirect rules "schedulers" or "schedulers/{name}"
 	r.URL.Path = strings.TrimRight(r.URL.Path, "/")
 	for _, rule := range h.microserviceRedirectRules {
+		// Now we only support checking the scheduling service whether it is independent
+		if rule.targetServiceName == mcsutils.SchedulingServiceName {
+			if !h.s.IsServiceIndependent(mcsutils.SchedulingServiceName) {
+				continue
+			}
+		}
 		if strings.HasPrefix(r.URL.Path, rule.matchPath) &&
 			slice.Contains(rule.matchMethods, r.Method) {
 			if rule.filter != nil && !rule.filter(r) {
 				continue
 			}
-			// we check the service primary addr here, so no need to check independently again.
+			// we check the service primary addr here,
+			// if the service is not available, we will return ErrRedirect by returning an empty addr.
 			addr, ok := h.s.GetServicePrimaryAddr(r.Context(), rule.targetServiceName)
 			if !ok || addr == "" {
 				log.Warn("failed to get the service primary addr when trying to match redirect rules",
 					zap.String("path", r.URL.Path))
+				return true, ""
 			}
 			// If the URL contains escaped characters, use RawPath instead of Path
 			origin := r.URL.Path

--- a/server/config/persist_options.go
+++ b/server/config/persist_options.go
@@ -608,7 +608,7 @@ func (o *PersistOptions) IsLocationReplacementEnabled() bool {
 	return o.getTTLBoolOr(sc.EnableLocationReplacement, o.GetScheduleConfig().EnableLocationReplacement)
 }
 
-// IsTikvRegionSplitEnabled returns whether tikv split region is disabled.
+// IsTikvRegionSplitEnabled returns whether tikv split region is enabled.
 func (o *PersistOptions) IsTikvRegionSplitEnabled() bool {
 	return o.getTTLBoolOr(sc.EnableTiKVSplitRegion, o.GetScheduleConfig().EnableTiKVSplitRegion)
 }

--- a/tests/integrations/mcs/tso/api_test.go
+++ b/tests/integrations/mcs/tso/api_test.go
@@ -18,6 +18,7 @@ import (
 	"bytes"
 	"context"
 	"encoding/json"
+	"fmt"
 	"io"
 	"net/http"
 	"testing"
@@ -199,4 +200,47 @@ func TestTSOServerStartFirst(t *testing.T) {
 	re.Len(group.Members, 2)
 
 	re.NoError(failpoint.Disable("github.com/tikv/pd/server/delayStartServerLoop"))
+}
+
+func TestForwardOnlyTSONoScheduling(t *testing.T) {
+	re := require.New(t)
+	re.NoError(failpoint.Enable("github.com/tikv/pd/pkg/utils/apiutil/serverapi/checkHeader", "return(true)"))
+	defer func() {
+		re.NoError(failpoint.Disable("github.com/tikv/pd/pkg/utils/apiutil/serverapi/checkHeader"))
+	}()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	tc, err := tests.NewTestAPICluster(ctx, 1)
+	defer tc.Destroy()
+	re.NoError(err)
+	err = tc.RunInitialServers()
+	re.NoError(err)
+	pdAddr := tc.GetConfig().GetClientURL()
+	ttc, err := tests.NewTestTSOCluster(ctx, 2, pdAddr)
+	re.NoError(err)
+	tc.WaitLeader()
+	leaderServer := tc.GetLeaderServer()
+	re.NoError(leaderServer.BootstrapCluster())
+
+	urlPrefix := fmt.Sprintf("%s/pd/api/v1", pdAddr)
+
+	// Test /operators, it should not forward when there is no scheduling server.
+	var slice []string
+	err = testutil.ReadGetJSON(re, dialClient, fmt.Sprintf("%s/%s", urlPrefix, "operators"), &slice,
+		testutil.WithoutHeader(re, apiutil.ForwardToMicroServiceHeader))
+	re.NoError(err)
+	re.Len(slice, 0)
+
+	// Test admin/reset-ts, it should forward to tso server.
+	input := []byte(`{"tso":"121312", "force-use-larger":true}`)
+	err = testutil.CheckPostJSON(dialClient, fmt.Sprintf("%s/%s", urlPrefix, "admin/reset-ts"), input,
+		testutil.StatusOK(re), testutil.StringContain(re, "Reset ts successfully"), testutil.WithHeader(re, apiutil.ForwardToMicroServiceHeader, "true"))
+	re.NoError(err)
+
+	// If close tso server, it should try forward to tso server, but return error in api mode.
+	ttc.Destroy()
+	err = testutil.CheckPostJSON(dialClient, fmt.Sprintf("%s/%s", urlPrefix, "admin/reset-ts"), input,
+		testutil.Status(re, http.StatusInternalServerError), testutil.StringContain(re, "[PD:apiutil:ErrRedirect]redirect failed"))
+	re.NoError(err)
 }

--- a/tests/server/api/scheduler_test.go
+++ b/tests/server/api/scheduler_test.go
@@ -26,7 +26,6 @@ import (
 
 	"github.com/pingcap/failpoint"
 	"github.com/pingcap/kvproto/pkg/metapb"
-	"github.com/stretchr/testify/require"
 	"github.com/stretchr/testify/suite"
 	sc "github.com/tikv/pd/pkg/schedule/config"
 	"github.com/tikv/pd/pkg/slice"
@@ -82,7 +81,7 @@ func (suite *scheduleTestSuite) checkOriginAPI(cluster *tests.TestCluster) {
 	re := suite.Require()
 	suite.NoError(tu.CheckPostJSON(testDialClient, urlPrefix, body, tu.StatusOK(re)))
 
-	suite.assertSchedulerExists(re, urlPrefix, "evict-leader-scheduler")
+	suite.assertSchedulerExists(urlPrefix, "evict-leader-scheduler")
 	resp := make(map[string]interface{})
 	listURL := fmt.Sprintf("%s%s%s/%s/list", leaderAddr, apiPrefix, server.SchedulerConfigHandlerPath, "evict-leader-scheduler")
 	suite.NoError(tu.ReadGetJSON(re, testDialClient, listURL, &resp))
@@ -94,20 +93,20 @@ func (suite *scheduleTestSuite) checkOriginAPI(cluster *tests.TestCluster) {
 	suite.NoError(err)
 	suite.NoError(failpoint.Enable("github.com/tikv/pd/pkg/schedule/schedulers/persistFail", "return(true)"))
 	suite.NoError(tu.CheckPostJSON(testDialClient, urlPrefix, body, tu.StatusNotOK(re)))
-	suite.assertSchedulerExists(re, urlPrefix, "evict-leader-scheduler")
+	suite.assertSchedulerExists(urlPrefix, "evict-leader-scheduler")
 	resp = make(map[string]interface{})
 	suite.NoError(tu.ReadGetJSON(re, testDialClient, listURL, &resp))
 	suite.Len(resp["store-id-ranges"], 1)
 	suite.NoError(failpoint.Disable("github.com/tikv/pd/pkg/schedule/schedulers/persistFail"))
 	suite.NoError(tu.CheckPostJSON(testDialClient, urlPrefix, body, tu.StatusOK(re)))
-	suite.assertSchedulerExists(re, urlPrefix, "evict-leader-scheduler")
+	suite.assertSchedulerExists(urlPrefix, "evict-leader-scheduler")
 	resp = make(map[string]interface{})
 	suite.NoError(tu.ReadGetJSON(re, testDialClient, listURL, &resp))
 	suite.Len(resp["store-id-ranges"], 2)
 	deleteURL := fmt.Sprintf("%s/%s", urlPrefix, "evict-leader-scheduler-1")
 	err = tu.CheckDelete(testDialClient, deleteURL, tu.StatusOK(re))
 	suite.NoError(err)
-	suite.assertSchedulerExists(re, urlPrefix, "evict-leader-scheduler")
+	suite.assertSchedulerExists(urlPrefix, "evict-leader-scheduler")
 	resp1 := make(map[string]interface{})
 	suite.NoError(tu.ReadGetJSON(re, testDialClient, listURL, &resp1))
 	suite.Len(resp1["store-id-ranges"], 1)
@@ -115,11 +114,11 @@ func (suite *scheduleTestSuite) checkOriginAPI(cluster *tests.TestCluster) {
 	suite.NoError(failpoint.Enable("github.com/tikv/pd/server/config/persistFail", "return(true)"))
 	err = tu.CheckDelete(testDialClient, deleteURL, tu.Status(re, http.StatusInternalServerError))
 	suite.NoError(err)
-	suite.assertSchedulerExists(re, urlPrefix, "evict-leader-scheduler")
+	suite.assertSchedulerExists(urlPrefix, "evict-leader-scheduler")
 	suite.NoError(failpoint.Disable("github.com/tikv/pd/server/config/persistFail"))
 	err = tu.CheckDelete(testDialClient, deleteURL, tu.StatusOK(re))
 	suite.NoError(err)
-	suite.assertNoScheduler(re, urlPrefix, "evict-leader-scheduler")
+	suite.assertNoScheduler(urlPrefix, "evict-leader-scheduler")
 	suite.NoError(tu.CheckGetJSON(testDialClient, listURL, nil, tu.Status(re, http.StatusNotFound)))
 	err = tu.CheckDelete(testDialClient, deleteURL, tu.Status(re, http.StatusNotFound))
 	suite.NoError(err)
@@ -492,7 +491,7 @@ func (suite *scheduleTestSuite) checkAPI(cluster *tests.TestCluster) {
 			testCase.extraTestFunc(testCase.createdName)
 		}
 		suite.deleteScheduler(urlPrefix, testCase.createdName)
-		suite.assertNoScheduler(re, urlPrefix, testCase.createdName)
+		suite.assertNoScheduler(urlPrefix, testCase.createdName)
 	}
 
 	// test pause and resume all schedulers.
@@ -507,7 +506,7 @@ func (suite *scheduleTestSuite) checkAPI(cluster *tests.TestCluster) {
 		body, err := json.Marshal(input)
 		suite.NoError(err)
 		suite.addScheduler(urlPrefix, body)
-		suite.assertSchedulerExists(re, urlPrefix, testCase.createdName) // wait for scheduler to be synced.
+		suite.assertSchedulerExists(urlPrefix, testCase.createdName) // wait for scheduler to be synced.
 		if testCase.extraTestFunc != nil {
 			testCase.extraTestFunc(testCase.createdName)
 		}
@@ -571,7 +570,7 @@ func (suite *scheduleTestSuite) checkAPI(cluster *tests.TestCluster) {
 			createdName = testCase.name
 		}
 		suite.deleteScheduler(urlPrefix, createdName)
-		suite.assertNoScheduler(re, urlPrefix, createdName)
+		suite.assertNoScheduler(urlPrefix, createdName)
 	}
 }
 
@@ -612,8 +611,8 @@ func (suite *scheduleTestSuite) checkDisable(cluster *tests.TestCluster) {
 	err = tu.CheckPostJSON(testDialClient, u, body, tu.StatusOK(re))
 	suite.NoError(err)
 
-	suite.assertNoScheduler(re, urlPrefix, name)
-	suite.assertSchedulerExists(re, fmt.Sprintf("%s?status=disabled", urlPrefix), name)
+	suite.assertNoScheduler(urlPrefix, name)
+	suite.assertSchedulerExists(fmt.Sprintf("%s?status=disabled", urlPrefix), name)
 
 	// reset schedule config
 	scheduleConfig.Schedulers = originSchedulers
@@ -623,7 +622,7 @@ func (suite *scheduleTestSuite) checkDisable(cluster *tests.TestCluster) {
 	suite.NoError(err)
 
 	suite.deleteScheduler(urlPrefix, name)
-	suite.assertNoScheduler(re, urlPrefix, name)
+	suite.assertNoScheduler(urlPrefix, name)
 }
 
 func (suite *scheduleTestSuite) addScheduler(urlPrefix string, body []byte) {
@@ -648,7 +647,7 @@ func (suite *scheduleTestSuite) testPauseOrResume(urlPrefix string, name, create
 		err := tu.CheckPostJSON(testDialClient, urlPrefix, body, tu.StatusOK(re))
 		re.NoError(err)
 	}
-	suite.assertSchedulerExists(re, urlPrefix, createdName) // wait for scheduler to be synced.
+	suite.assertSchedulerExists(urlPrefix, createdName) // wait for scheduler to be synced.
 
 	// test pause.
 	input := make(map[string]interface{})
@@ -727,25 +726,33 @@ func (suite *scheduleTestSuite) checkEmptySchedulers(cluster *tests.TestCluster)
 	}
 }
 
-func (suite *scheduleTestSuite) assertSchedulerExists(re *require.Assertions, urlPrefix string, scheduler string) {
+func (suite *scheduleTestSuite) assertSchedulerExists(urlPrefix string, scheduler string) {
 	var schedulers []string
+	re := suite.Require()
 	tu.Eventually(re, func() bool {
-		tu.ReadGetJSON(suite.Require(), testDialClient, urlPrefix, &schedulers)
+		err := tu.ReadGetJSON(re, testDialClient, urlPrefix, &schedulers,
+			tu.StatusOK(re))
+		suite.NoError(err)
 		return slice.Contains(schedulers, scheduler)
 	})
 }
 
-func (suite *scheduleTestSuite) assertNoScheduler(re *require.Assertions, urlPrefix string, scheduler string) {
+func (suite *scheduleTestSuite) assertNoScheduler(urlPrefix string, scheduler string) {
 	var schedulers []string
+	re := suite.Require()
 	tu.Eventually(re, func() bool {
-		tu.ReadGetJSON(suite.Require(), testDialClient, urlPrefix, &schedulers)
+		err := tu.ReadGetJSON(re, testDialClient, urlPrefix, &schedulers,
+			tu.StatusOK(re))
+		suite.NoError(err)
 		return !slice.Contains(schedulers, scheduler)
 	})
 }
 
 func (suite *scheduleTestSuite) isSchedulerPaused(urlPrefix, name string) bool {
 	var schedulers []string
-	err := tu.ReadGetJSON(suite.Require(), testDialClient, fmt.Sprintf("%s?status=paused", urlPrefix), &schedulers)
+	re := suite.Require()
+	err := tu.ReadGetJSON(re, testDialClient, fmt.Sprintf("%s?status=paused", urlPrefix), &schedulers,
+		tu.StatusOK(re))
 	suite.NoError(err)
 	for _, scheduler := range schedulers {
 		if scheduler == name {

